onnx/defs/tensor/defs.cc

```diff
@@ -77,19 +77,126 @@ could also be 0, in which case the actual dimension value is unchanged (i.e. tak
 from the input tensor).)DOC";
 
 ONNX_OPERATOR_SET_SCHEMA(
-    Reshape,
-    5,
+    Reshape, 5,
     OpSchema()
         .SetDoc(Reshape_ver5_doc)
         .Input(0, "data", "An input tensor.", "T")
         .Input(1, "shape", "Specified shape for output.", "tensor(int64)")
         .Output(0, "reshaped", "Reshaped data.", "T")
-        .TypeConstraint(
-            "T",
-            {"tensor(float16)", "tensor(float)", "tensor(double)"},
-            "Constrain input and output types to float tensors.")
-        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+        .TypeConstraint("T",
+                        {"tensor(float16)", "tensor(float)", "tensor(double)"},
+                        "Constrain input and output types to float tensors.")
+        .TypeAndShapeInferenceFunction([](InferenceContext &ctx) {
+          // Type Inference
           propagateElemTypeFromInputToOutput(ctx, 0, 0);
+          // Shape Inference if target shape is in initializer
+          const TensorProto *targetShapeInitializer =
+              ctx.getInputInitializer(1);
+          if (!targetShapeInitializer) {
+            return;
+          }
+          // Make targetShape (0 -> same as originalShape, -1 -> inferred).
+          // The targetShape vector represents the specified shape for output.
+          std::vector<int64_t> targetShape;
+          if (targetShapeInitializer->has_raw_data()) {
+            const std::string &bytes = targetShapeInitializer->raw_data();
+            targetShape.insert(targetShape.end(),
+                               reinterpret_cast<const int64_t *>(bytes.c_str()),
+                               reinterpret_cast<const int64_t *>(bytes.c_str() +
+                                                                 bytes.size()));
+          } else {
+            const auto &data = targetShapeInitializer->int64_data();
+            targetShape.insert(targetShape.end(), data.begin(), data.end());
+          }
+
+          // Iterate through targetShape, adding dimensions in the outputShape
+          // TensorProto. If the targertShape dimension is -1, we do not set the
```

Comment:
Should we support the case that 2nd input is available and its value is -1? if yes, what's the scenario please?

Reply:
@linkerzhang Could you clarify by "its value is -1" you mean the second input is a scalar value `-1` or the second input is a 1D tensor which has an item being `-1`? The latter is supported in @varunjain99's current implementation, what is he saying in this comment is the whole inference process has two stages, and this first stage will mark the `-1` axis, and then compute it in the second stage (see negativeOneDim).

Reply:
I mean the latter one.

I was thinking whether we should support the latter case in this shape inference function. It depends on when we want to run this shape inference function,  if it's before the data coming only, then no need to implement this case here I think. If we also run this function after we get the input data, it's fine to have it here.

Reply:
@linkerzhang Why are there differences between with and without real input?

Reply:
The purpose of this function is that if the reshape shape (data of input 2) is present, we try to infer as much of the output shape. When there is a -1 in the 2nd input data, we can try to infer this dimension if we have enough *shape* information from input 1. We don't need the data of input 1, but enough shape information from input 1, to figure out what -1 should actually be.